File: leetcode/2521. Distinct Prime Factors of Product of Array/2521.py
# https://helloacm.com/teaching-kids-programming-distinct-prime-factors-of-product-of-array/
# https://leetcode.com/problems/distinct-prime-factors-of-product-of-array/
# MEDIUM, MATH

class Solution:
    def distinctPrimeFactors(self, nums: List[int]) -> int:
        a = set()
        for x in nums:
            j = 2
            while x != 1:
                while x % j == 0:
                    x //= j
                    a.add(j)
                j += 1
        return len(a)

File: leetcode/2521. Distinct Prime Factors of Product of Array/2521.rs
// https://helloacm.com/teaching-kids-programming-distinct-prime-factors-of-product-of-array/
// https://leetcode.com/problems/distinct-prime-factors-of-product-of-array/
// MEDIUM, MATH

impl Solution {
    pub fn distinct_prime_factors(mut nums: Vec<i32>) -> i32 {
        use std::collections::HashSet;
        let mut a = HashSet::new();
        for x in nums.iter_mut() {
            let mut p = 2;
            while (*x != 1) {
                while *x % p == 0 {
                    *x /= p;
                    a.insert(p);
                }
                p += 1;
            }
        }
        return a.len() as _;
    }
}
